src/client.rs

```diff
@@ -403,166 +431,184 @@ impl Spotify {
         market: Option<Country>,
         include_external: Option<IncludeExternal>,
     ) -> ClientResult<SearchResult> {
-        let mut params = HashMap::new();
-        let limit = limit.into().unwrap_or(10);
-        let offset = offset.into().unwrap_or(0);
-        if let Some(_market) = market {
-            params.insert("market".to_owned(), _market.as_str().to_owned());
+        let mut params = Query::with_capacity(4);
+        params.insert("limit".to_owned(), limit.into().unwrap_or(10).to_string());
+        params.insert("offset".to_owned(), offset.into().unwrap_or(0).to_string());
+        params.insert("q".to_owned(), q.to_owned());
+        params.insert("type".to_owned(), _type.as_str().to_owned());
+        if let Some(market) = market {
```

Comment:
I did think about it. I initially wanted to use a `serde_json::Value` type for GET requests too because it makes its usage super easy and clean, but I ended up running into too many issues.

I also tried to make the query type a `HashMap<&str, &str>` but failed because of ownership issues (this would avoid the `to_owned` calls). All of the following conditional inserts failed because `&str` is not owned and it goes out of scope:

```rust
        if let Some(market) = market {
            params.insert("market", market.as_str());
        }
```

But I can give that a try again to see if it's possible to improve before being forced to use a macro (you can give it a try as well). Rather than creating one I'd use existing crates, like [maplit](https://github.com/bluss/maplit).

Reply:
I would leave this for a different PR, though.